### PR TITLE
Add queryable interfaces

### PR DESCRIPTION
**Pull request by rubensworks**
This PR implements the latest state of the new query spec as defined in https://github.com/rdfjs/query-spec/

This PR should not be merged yet into master until the query spec is stable.
In the meantime, prereleases of this package are being published to npm for testing purposes.
At the time of writing `1.1.0-next.0` of `@rdfjs/types` is available on npm.

@jacoscaz I changed some small things while experimenting with these interfaces in Comunica. Nothing major was changed, just some things to simplify the implementation of them.
All interfaces (except for the filterable ones) have been implemented in Comunica, and have been confirmed to work as intended.

### CODE REVIEW

File: query/common.d.ts
Comment (jacoscaz) [inline query/common.d.ts:203]: Moving modifiers at the instance level is something I had to do, too. +1 from me.

**Review by jacoscaz — COMMENTED** [review]
Excellent work @rubensworks ! I'm really happy to hear that everything seems to work as intended. I have a few comments but only one of them is significant.

File: query/common.d.ts
Comment (jacoscaz) [reply query/common.d.ts:18]: Reference: https://github.com/rdfjs/types/blob/b04a4158c7dabbcd487661255afbe7c9eac5f25f/stream.d.ts#L22

Comment (jacoscaz) [reply query/common.d.ts:18]: We could have a `Stream<Q extends BaseQuad | Bindings>`, a `QuadStream<Q extends BaseQuad = Quad> extends Stream<Q>` and a `BindingsStream extends Stream<Bindings>`

Comment (rubensworks) [reply query/common.d.ts:18]: Yes, that's possible indeed. But the problem with that is that this will imply a breaking change in the typings, which might not be pleasant for downstream users.
Perhaps we can add a TODO to merge them from the moment we require a breaking change?

File: query/common.d.ts
Comment (rubensworks) [reply query/common.d.ts:97]: Right, good point. Will have a look at that.

Comment (rubensworks) [reply query/common.d.ts:105]: Ok, pushed some changes (and tested in Comunica). Can you check if you agree with those?

File: query/common.d.ts
Comment (jacoscaz) [reply query/common.d.ts:18]: True! Yes, I think a TODO would be a good idea.

File: query/common.d.ts
Comment (jacoscaz) [reply query/common.d.ts:105]: Oh man, why did I not think of limiting the options myself? I went for an insanely complicated approach... Oh well. Nice!

Comment (rubensworks) [reply query/common.d.ts:105]: I was just following your exact instructions, so the solution is all yours 😅

**Review by jacoscaz — APPROVED** [review]
Looks good to me! I wouldn't merge, yet, but looks good.

**Comment by rubensworks** [comment]
No indeed, won't merge yet. Just published `1.1.0-next.1` for whoever is interested.

**Comment by rubensworks** [comment]
I've removed the filterable interfaces from this PR for now into #31, since no implementations of this interface exist yet, and it may take a longer time to stabilize.

This PR has been open for a week now.
Unless there are any objections, I'll wait another week (until February 17) to merge this.

File: query/queryable.d.ts
Comment (tpluscode) [inline query/queryable.d.ts:73]: What is the use case for a non-SPARQL query engine?

**Review by tpluscode — COMMENTED** [review]
I realize I come late to the party, since I was not actively participating in this PR. Sorry if you might find my comment a little disruptive of your effort.

I envisioned that I would have an immediate use for it but I gotta share some observations. My biggest beef is with the `Bindings` interface. Please see the individual code comments

File: query/queryable.d.ts
Comment (rubensworks) [reply query/queryable.d.ts:139]: There was also a long discussion about this one here: https://github.com/rdfjs/query-spec/pull/7

Eventually, we reached a consensus about the current methods.
(I don't remember the exact argumentation, perhaps @jacoscaz or @gsvarovsky knows?)

File: query/queryable.d.ts
Comment (rubensworks) [reply query/queryable.d.ts:73]: GraphQL(-LD) for instance.

File: query/queryable.d.ts
Comment (rubensworks) [reply query/queryable.d.ts:22]: > Doesn't the use of any make this whole interface effectively Record<string, any>?

Yes, we went forth and back on this one. Not sure anymore why we eventually landed on this one. Do you remember @jacoscaz?

> Maybe we could suggest extending there interfaces via [module augmentation](https://www.typescriptlang.org/docs/handbook/declaration-merging.html#module-augmentation)?

Not sure I understand. Could you give an example where module augmentation could be relevant?

Comment (tpluscode) [reply query/queryable.d.ts:22]: Ok my mistake, I think I underestimated TS or maybe something changed in recent versions.

The indexer is not an issue per-se apparently but I would in turn ask what is its purpose? Is the intention is to allow consumers to extend the context interface?

With augmentation they would write

```ts
declare module '@rdfjs/types' {
  interface QueryContext<SourceType> { // not sure if the generic arg needs to be repeated
    foo?: string
    bar?: Baz
  }
}
```

**Review by tpluscode — COMMENTED** [review]
Closing thought: for a large PR with many new pieces, it would be wise to add some tests :)

File: .changeset/pre.json
Comment (tpluscode) [inline .changeset/pre.json:1]: Aha! This will be the real last comment. Should the `pre` take be removed or are you intending another prerelease?

File: query/queryable.d.ts
Comment (jacoscaz) [reply query/queryable.d.ts:69]:
```typescript
type SomeUnknownType = unknown;
type IsThisUnknownToo = SomeUnknownType | string;

type SomeAnyType = any;
type IsThisAnyToo = SomeAnyType | string;
```

These seem to both collapse into `unknown` and `any`, respectively.

**Comment by rubensworks** [comment]
> Closing thought: for a large PR with many new pieces, it would be wise to add some tests :)

Right, will look into that.

File: .changeset/pre.json
Comment (rubensworks) [reply .changeset/pre.json:1]: Yes, will remove.

File: query/queryable.d.ts
Comment (tpluscode) [inline query/queryable.d.ts:105]:
```suggestion
 * SPARQL-constrained query interface.
```

just noticed a typo

File: query/queryable.d.ts
Comment (gsvarovsky) [reply query/queryable.d.ts:139]: I'm definitely in favour of more fluent method names. My only reservation is that these methods obtain a query object/proxy but don't execute it. But at this point that's probably splitting hairs.

Comment (tpluscode) [reply query/queryable.d.ts:139]: What does it mean that they return a proxy?

Comment (jacoscaz) [reply query/queryable.d.ts:139]: If we all agree on having these methods return results directly, without an intermediate `Query` object, then we might also agree on changing their names as per @tpluscode's suggestion. Anyone against this?

File: query/common.d.ts
Comment (jacoscaz) [reply query/common.d.ts:184]: >> it merely gives an illusion of type safety rather than actual type safety

> This pretty much sums up all of TypeScript, doesn't it? :)

Well, yes and no. TypeScript disappears at runtime and I agree it's not enough on its own to *guarantee* type safety if a program does any kind of input/output. However, one can use different strategies to ensure some level of safety. Reading from a relational database, for example, provides for some level of safety as long as the types are kept in sync with the schema, either manually or with some automation. If no schema is guaranteed at the source level we can always use validation.

However, a JS spec working on graphs where the only thing that can be taken for granted is the `S P O (G)` structure offers very little in terms of type safety and I fear that providing an easy/implicit way to cast query results into specific shapes might encourage a misplaced sense of security.

All that said, I do understand the argument for convenience. What if we were to offer a dedicated `TypedBindings` interfaces that extends the untype one, leaving the responsibility of casting one into the other to users?

```typescript
import { Bindings } from './query';
import { Variable, Term } from './data-model';

// This would have to become the new `Variable`. Setting `Value` to `string`
// by default should make this a non-breaking change.
interface TypedVariable<Value extends string = string> extends Variable {
  value: Value;
}

interface TypedBindings<Shape extends { [key: string]: Term }> extends Bindings {
  get<Key extends Extract<keyof Shape, string>>(key: Key): Shape[Key];
  get<Key extends Extract<keyof Shape, string>>(variable: TypedVariable<Key>): Shape[Key];
}
```

With this, one could easily do `(someBindingsObj as TypedBindings<SomeShape>)` . As per your comment,  it would require making `Variable` generic.

File: query/queryable.d.ts
Comment (rubensworks) [reply query/queryable.d.ts:139]: I just remembered a (significant) reason to keep the `query...` methods: interface discovery through autocompletion.

Since the `Queryable` interface already exposes a `query` method, developers that don't need the API docs might start typing `query...` in their editor. Smart editors would then immediately also suggest `queryBindings`, `queryQuads`, ... This would help developers get directed to the simpler query methods from the `SparqlQueryable` interface _if_ they are available.

If we would go for `select`, `construct`, etc, then we have to take into account the fact that `construct` might cause some confusion, since we don't have a 1-1 mapping with the SPARQL query clauses. Namely, `construct` would be used for both `CONSTRUCT` and `DESCRIBE` query clauses, which may not be obvious for newcomers.

File: query/common.d.ts
Comment (rubensworks) [reply query/common.d.ts:184]: It seems to me like quite some (discussion) work will have to go into defining such typed bindings. In any case, this would result in a breaking change for the current typings, so we probably don't want to do this in the near future. Given this, shall we move this discussion to a new issue?

In any case, I will look into adding a string-based `get` method on `Bindings`, as I agree that it might be more convenient in some cases.

File: query/queryable.d.ts
Comment (jacoscaz) [reply query/queryable.d.ts:139]: Well, if were to go with renaming to `select`, `construct`, ...  then we should also have a `describe`, which in the majority of cases would probably call `construct` internally. I guess on one side we would favor discovery by autocompletion, on the other side we would favor discovery by intuition. I can see the point in both approaches, happy to go with either.

File: query/common.d.ts
Comment (jacoscaz) [reply query/common.d.ts:184]: I agree with filing this into a separate issue. Typed bindings may take a lot of discussion indeed.

File: query/queryable.d.ts
Comment (rubensworks) [reply query/queryable.d.ts:139]: Another (minor) argument in favor of `query...` is the alignment with the `resultType` field of `Query`.

**Comment by rubensworks** [comment]
I just added some tests (I didn't aim for full coverage though, just the most critical parts for now)

**Comment by rubensworks** [comment]
All issues should be resolved now. Unless there are any other comments, I'll merge this on Thursday (after fixing the changelog).

File: query/queryable.d.ts
Comment (jacoscaz) [reply query/queryable.d.ts]: All told, there seem to be more arguments in favor of keeping these methods as they are. I think we can resolve this and proceed with merging.

**Pull request merged**
